swift/ql/lib/codeql/swift/frameworks/StandardLibrary/CustomUrlSchemes.qll

```diff
@@ -56,3 +62,71 @@ private class LaunchOptionsUrlVarDecl extends VarDecl {
     this.getName() = "url"
   }
 }
+
+/** A type or extension declaration adopting the protocol `UISceneDelegate`. */
+private class AdoptingUiSceneDelegate extends Decl {
+  AdoptingUiSceneDelegate() {
+    exists(ProtocolDecl delegate |
+      this.(ExtensionDecl).getAProtocol().getABaseTypeDecl*() = delegate or
+      this.(ClassOrStructDecl).getABaseTypeDecl*() = delegate
+    |
+      delegate.getName() = "UISceneDelegate"
+    )
+  }
+}
+
+/**
+ * An `OpenURLContexts`, `NSUserActivity`, or `ConnectionOptions` parameter of a `scene` method
+ * declared in a type adopting `UISceneDelegate`.
+ */
+// This is a temporary workaround until the TODO 2 above is addressed.
+private class UiSceneDelegateSource extends RemoteFlowSource {
+  UiSceneDelegateSource() {
+    exists(FuncDecl f, ParamDecl p, AdoptingUiSceneDelegate d |
+      f.getName() = "scene(_:" + ["continue", "didUpdate", "openURLContexts"] + ":)" and
+      p = f.getParam(1)
+      or
+      f.getName() = "scene(_:willConnectTo:options:)" and
+      p = f.getParam(2)
+    |
+      f.getEnclosingDecl() = d and
+      this.(DataFlow::ParameterNode).getParameter() = p
+    )
+  }
+
+  override string getSourceType() { result = "Remote data in UIWindowSceneDelegate.scene" }
+}
+
+/**
+ * A content implying that, if a `UIOpenURLContext` is tainted, then its field `url` is also tainted.
+ */
+private class UiOpenUrlContextUrlInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent {
+  UiOpenUrlContextUrlInheritTaint() {
+    this.getField().getEnclosingDecl().(NominalTypeDecl).getName() = "UIOpenURLContext" and
+    this.getField().getName() = "url"
+  }
+}
+
+/**
+ * A content implying that, if a `NSUserActivity` is tainted, then its field `webpageURL` is also tainted.
+ */
+private class UserActivityUrlInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent {
+  UserActivityUrlInheritTaint() {
+    this.getField().getEnclosingDecl().(NominalTypeDecl).getName() = "NSUserActivity" and
+    this.getField().getName() = "webpageURL"
+  }
+}
+
+/**
+ * A content implying that, if a `UIScene.ConnectionOptions` is tainted, then its fields
+ * `userActivities` and `urlContexts` are also tainted.
+ */
+private class ConnectionOptionsFieldsInheritTaint extends TaintInheritingContent,
+  DataFlow::Content::FieldContent {
+  ConnectionOptionsFieldsInheritTaint() {
+    this.getField().getEnclosingDecl().(NominalTypeDecl).getName() = "ConnectionOptions" and
```

Comment:
We do have `NominalTypeDecl.getFullName()` if we want to match the full name `UIScene.ConnectionOptions` here?

Reply:
Similarly to another case in another PR, using `getFullName()` works in tests but fails in a real codebase for certain types. I've opened an internal issue for it.

Reply:
Thanks.